src/psyclone/psyGen.py

```diff
@@ -3083,6 +3083,129 @@ def scalars(self):
             "Arguments.scalars must be implemented in sub-class")
 
 
+class DataAccess(object):
+    '''A helper class to handle dependencies due to overlapping data
+    accesses between fields
+
+    '''
+
+    def __init__(self, arg):
+        '''store the source argument and its parent call. Also set any
+        internal coverage data.
+
+        :param arg: the argument that we are concerned with
+        :type arg: :py:class:`psyclone.psyGen.Argument`
+
+        '''
+        self._arg = arg
+        self._call = arg.call
+        self.reset_coverage()
+
+    def overlaps(self, arg):
+        '''Determine whether the accesses to the provided field overlap
+        with the accesses of the source argument
+
+        We do not currently deal with accesses to a subset of a field
+        (unless it is a vector). This distinction will need to be
+        added once loop splitting is supported and would be useful for
+        boundary condition kernels
+
+        :param arg: the argument to compare with our internal argument
+        :type arg: :py:class:`psyclone.psyGen.Argument`
+        :return Bool: True is there are overlapping accesses between \
+                      arguments and False if not.
+
+        '''
+        if self._arg.name != arg.name:
+            # the arguments are different args so do not overlap
+            return False
+
+        if isinstance(self._call, HaloExchange) and \
+           isinstance(arg.call, HaloExchange) and \
+           (self._arg.vector_size > 1 or arg.vector_size > 1):
+            # This is a vector field and both accesses come from halo
+            # exchanges. As halo exchanges only access a particular
+            # vector the accesses do not overlap if the vector indices
+            # being accessed differ.
+
+            # sanity check
+            self._call.check_vector_halos_differ(arg.call)
+
+            if self._call.vector_index != arg.call.vector_index:
+                # accesses are to different vector indices so do not overlap
+                return False
+        # accesses do overlap
+        return True
+
+    def reset_coverage(self):
+        '''Reset internal state to allow re-use of the object for a different
+        situation.
+
+        '''
+        self._covered = False
+        self._vector_index_access = []
+
+    def update_coverage(self, arg):
+        '''Record any overlap between accesses to the supplied argument and
+        the source argument. If the overlap results in all of the
+        accesses to the source argument being covered (either directly
```

Comment:
Again, please re-word this to explain what's going on - i.e. what does "being covered" mean?

Reply:
I've tred.